pkg/tracking/store/sql/runs_search_builder.go

```diff
@@ -0,0 +1,489 @@
+package sql
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+
+	"github.com/mlflow/mlflow-go/pkg/contract"
+	"github.com/mlflow/mlflow-go/pkg/protos"
+	"github.com/mlflow/mlflow-go/pkg/tracking/service/query"
+	"github.com/mlflow/mlflow-go/pkg/tracking/service/query/parser"
+	"github.com/mlflow/mlflow-go/pkg/tracking/store/sql/models"
+	"github.com/mlflow/mlflow-go/pkg/utils"
+)
+
+func applyLifecycleStagesFilter(runViewType protos.ViewType) []models.LifecycleStage {
+	switch runViewType {
+	case protos.ViewType_ACTIVE_ONLY:
+		return []models.LifecycleStage{
+			models.LifecycleStageActive,
+		}
+	case protos.ViewType_DELETED_ONLY:
+		return []models.LifecycleStage{
+			models.LifecycleStageDeleted,
+		}
+	case protos.ViewType_ALL:
+		return []models.LifecycleStage{
+			models.LifecycleStageActive,
+			models.LifecycleStageDeleted,
+		}
+	}
+
+	return []models.LifecycleStage{
+		models.LifecycleStageActive,
+		models.LifecycleStageDeleted,
+	}
+}
+
+func getOffset(pageToken string) (int, *contract.Error) {
+	if pageToken != "" {
+		var token PageToken
+		if err := json.NewDecoder(
+			base64.NewDecoder(
+				base64.StdEncoding,
+				strings.NewReader(pageToken),
+			),
+		).Decode(&token); err != nil {
+			return 0, contract.NewErrorWith(
+				protos.ErrorCode_INVALID_PARAMETER_VALUE,
+				fmt.Sprintf("invalid page_token: %q", pageToken),
+				err,
+			)
+		}
+
+		return int(token.Offset), nil
+	}
+
+	return 0, nil
+}
+
+//nolint:funlen,cyclop,gocognit
+func applyFilter(ctx context.Context, database, transaction *gorm.DB, filter string) *contract.Error {
```

Comment:
I don't mind the `nolint` here, this is reasonable. You are just dealing with all the sql stuff in this function, it is hard to keep this short and readable.